### DISCUSSION TITLE
Rplidar A1M8 connection on PIO pins

### DISCUSSION THREAD

**Discussion by alain-lnpcrd** [iRobotEducation/create3_docs#520] [Questions]
### How are you connecting to your Create 3?

Wi-Fi (ROS 2)

### Computer(s) Model(s) and Operating System(s)

RaspberryPI 4B 8Go

### Which version of ROS 2 is installed on your computer?

Humble

### Which firmware version is installed on your robot?

H.2.3

### Which RMW is your robot running?

FastRTPS

### Does your robot have an assigned namespace? If so, please share.

/

### Is the robot connected to a network? If so, what is the network type?

home network

### Are there multiple Create 3 robots connected to your network?

No

### Is multicast enabled?

None

### What is the Adapter Board's USB/BLE Toggle currently switched to?

USB

### Describe your question.

Hi,
The USB/UART adapter of my rplidar being damaged and given that the only supplier found RobotShop is in stock renewal, I fell back on the PIO connection described on https://github.com/berndporr/rplidar_rpi.
The example
sudo ./printdata > tt2.tsv
gnuplot> plot "tt2.tsv"
corrected to:
sudo ./printdata > tt2.tsv
gnuplot
set terminal jpeg
set output "output.jpg"
plot "tt2.tsv"
gives with output.jpg the image of a laser trace.

But I can't get the mapping proposed in create3_lidar to work.
The serial_port having changed I create a symbolic link:
ln -s /dev/ttyAMA0 /dev/ttyUSB0
  then I run the launches

```
moi@pi4b-8go-u-22-04:~/create3_ws$ ros2 launch create3_lidar_slam sensors_launch.py
[INFO] [launch]: All log files can be found below /home/moi/.ros/log/2024-01-18-11-24-26-117298-pi4b-8go-u-22-04-4562
[INFO] [launch]: Default logging verbosity is set to INFO
[INFO] [static_transform_publisher-1]: process started with pid [4563]
[static_transform_publisher-1] [WARN] [1705573466.878842845] []: Old-style arguments are deprecated; see --help for new-style arguments
[static_transform_publisher-1] [INFO] [1705573467.033781699] [static_transform_publisher_fRacLjn3XmrW5rbJ]: Spinning until stopped - publishing transform
[static_transform_publisher-1] translation: ('-0.012000', '0.000000', '0.144000')
[static_transform_publisher-1] rotation: ('0.000000', '0.000000', '0.000000', '1.000000')
[static_transform_publisher-1] from 'base_footprint' to 'laser_frame'
[INFO] [rplidar_composition-2]: process started with pid [4574]
[rplidar_composition-2] [INFO] [1705573468.967938656] [rplidar_composition]: RPLidar running on ROS2 package rplidar_ros. RPLIDAR SDK Version:2.1.0

```
```
moi@pi4b-8go-u-22-04:~/create3_ws$ ros2 launch create3_lidar_slam slam_toolbox_launch.py 
[INFO] [launch]: All log files can be found below /home/moi/.ros/log/2024-01-18-11-09-09-077465-pi4b-8go-u-22-04-4384
[INFO] [launch]: Default logging verbosity is set to INFO
[INFO] [async_slam_toolbox_node-1]: process started with pid [4385]
[async_slam_toolbox_node-1] [INFO] [1705572549.954706633] [slam_toolbox]: Node using stack size 40000000
[async_slam_toolbox_node-1] [INFO] [1705572550.512325153] [slam_toolbox]: Using solver plugin solver_plugins::CeresSolver
[async_slam_toolbox_node-1] [INFO] [1705572550.516101922] [slam_toolbox]: CeresSolver: Using SCHUR_JACOBI preconditioner.

```
moi@pi4b-8go-u-22-04:~/create3_ws$ ros2 launch create3_lidar_slam rviz_launch.py 
[INFO] [launch]: All log files can be found below /home/moi/.ros/log/2024-01-18-11-09-36-475551-pi4b-8go-u-22-04-4413
[INFO] [launch]: Default logging verbosity is set to INFO
[INFO] [rviz2-1]: process started with pid [4414]
[rviz2-1] [INFO] [1705572580.075111606] [rviz2]: Stereo is NOT SUPPORTED
[rviz2-1] [INFO] [1705572580.075437472] [rviz2]: OpenGl version: 4.5 (GLSL 4.5)
[rviz2-1] [INFO] [1705572582.503774911] [rviz2]: Stereo is NOT SUPPORTED

```
 No mapping in RVIZ, frame [map] doesn't exist

moi@pi4b-8go-u-22-04:~/create3_ws$ ros2 node list
/_internal/composite_hazard
/_internal/kinematics_engine
/_internal/mobility
/_internal/stasis
/mobility_monitor
/motion_control
/robot_state
/rplidar_composition
/rviz
/slam_toolbox
/static_transform
/static_transform_publisher_fRacLjn3XmrW5rbJ
/system_health
/transform_listener_impl_aaaac04e7530
/transform_listener_impl_aaaae66aa630
/ui_mgr
moi@pi4b-8go-u-22-04:~/create3_ws$ 

No rplidar_node
No publisher

```
moi@pi4b-8go-u-22-04:~/create3_ws$ ros2 topic info --verbose /scan
Type: sensor_msgs/msg/LaserScan

Publisher count: 0

Subscription count: 2

Node name: slam_toolbox
Node namespace: /
Topic type: sensor_msgs/msg/LaserScan
Endpoint type: SUBSCRIPTION
GID: 01.0f.7d.2d.21.11.44.bf.02.00.00.00.00.00.22.04.00.00.00.00.00.00.00.00
QoS profile:
  Reliability: BEST_EFFORT
  History (Depth): UNKNOWN
  Durability: VOLATILE
  Lifespan: Infinite
  Deadline: Infinite
  Liveliness: AUTOMATIC
  Liveliness lease duration: Infinite

Node name: rviz
Node namespace: /
Topic type: sensor_msgs/msg/LaserScan
Endpoint type: SUBSCRIPTION
GID: 01.0f.7d.2d.3e.11.b6.a4.01.00.00.00.00.00.29.04.00.00.00.00.00.00.00.00
QoS profile:
  Reliability: RELIABLE
  History (Depth): UNKNOWN
  Durability: VOLATILE
  Lifespan: Infinite
  Deadline: Infinite
  Liveliness: AUTOMATIC
  Liveliness lease duration: Infinite

moi@pi4b-8go-u-22-04:~/create3_ws$ 

```
I probably have to touch the program but I have no idea where

But So,

```
moi@pi4b-8go-u-22-04:~/create3_ws$ ros2 launch rplidar_ros view_rplidar_a1_launch.py
[INFO] [launch]: All log files can be found below /home/moi/.ros/log/2024-01-18-13-10-34-186673-pi4b-8go-u-22-04-4786
[INFO] [launch]: Default logging verbosity is set to INFO
[INFO] [rplidar_node-1]: process started with pid [4787]
[INFO] [rviz2-2]: process started with pid [4789]
[rplidar_node-1] [INFO] [1705579834.858576144] [rplidar_node]: RPLidar running on ROS2 package rplidar_ros. RPLIDAR SDK Version:2.1.0
[rviz2-2] [INFO] [1705579838.237953699] [rviz2]: Stereo is NOT SUPPORTED
[rviz2-2] [INFO] [1705579838.238655431] [rviz2]: OpenGl version: 4.5 (GLSL 4.5)
[rviz2-2] [INFO] [1705579840.767183752] [rviz2]: Stereo is NOT SUPPORTED

```
Grid is empty and Frale [laser] does not exist


Thanks for help!

**Comment by alain-lnpcrd** [comment]
hi,
The port for GPIO 14/15 TXD/RXD on an RPi4 is /dev/ttyS0
I plugged the motor into the USB/UART which allows the rplidar to run

```
moi@pi4b-8go-u-22-04:~/create3_ws$ sudo ln -s /dev/ttyS0 /dev/ttyUSB0
[sudo] Mot de passe de moi : 
moi@pi4b-8go-u-22-04:~/create3_ws$ ros2 launch rplidar_ros view_rplidar_a1_launch.py
[INFO] [launch]: All log files can be found below /home/moi/.ros/log/2024-01-22-11-22-35-367271-pi4b-8go-u-22-04-3451
[INFO] [launch]: Default logging verbosity is set to INFO
serial_port <launch.substitutions.launch_configuration.LaunchConfiguration object at 0xffffb66921a0>
[INFO] [rplidar_node-1]: process started with pid [3453]
[INFO] [rviz2-2]: process started with pid [3455]
[rplidar_node-1] [INFO] [1705918956.031176233] [rplidar_node]: RPLidar running on ROS2 package rplidar_ros. RPLIDAR SDK Version:2.1.0
[rplidar_node-1] [INFO] [1705918956.038369349] [rplidar_node]: RPLidar S/N: 699E99F6C9E59AD5C5E59CF7954C3415
[rplidar_node-1] [INFO] [1705918956.038564865] [rplidar_node]: Firmware Ver: 1.29
[rplidar_node-1] [INFO] [1705918956.038605643] [rplidar_node]: Hardware Rev: 7
[rplidar_node-1] [INFO] [1705918956.041198685] [rplidar_node]: RPLidar health status : 0
[rplidar_node-1] [INFO] [1705918956.041335721] [rplidar_node]: RPLidar health status : OK.
[rplidar_node-1] [INFO] [1705918956.041368757] [rplidar_node]: Start
[rplidar_node-1] [INFO] [1705918956.236846090] [rplidar_node]: current scan mode: Sensitivity, sample rate: 8 Khz, max_distance: 12.0 m, scan frequency:10.0 Hz, 
[rviz2-2] [INFO] [1705918959.346897040] [rviz2]: Stereo is NOT SUPPORTED
[rviz2-2] [INFO] [1705918959.347467348] [rviz2]: OpenGl version: 4.5 (GLSL 4.5)
[rviz2-2] [INFO] [1705918962.089774962] [rviz2]: Stereo is NOT SUPPORTED

```
![image](https://github.com/iRobotEducation/create3_docs/assets/74408456/960be1f4-eded-4e02-81d5-3f0452474b44)

but
```
moi@pi4b-8go-u-22-04:~/create3_ws$ ros2 launch create3_lidar_slam sensors_launch.py 
[INFO] [launch]: All log files can be found below /home/moi/.ros/log/2024-01-22-11-44-11-910842-pi4b-8go-u-22-04-8173
[INFO] [launch]: Default logging verbosity is set to INFO
[INFO] [static_transform_publisher-1]: process started with pid [8174]
[static_transform_publisher-1] [WARN] [1705920252.498046882] []: Old-style arguments are deprecated; see --help for new-style arguments
[static_transform_publisher-1] [INFO] [1705920252.560378532] [static_transform_publisher_Pci896eJelZdokVR]: Spinning until stopped - publishing transform
[static_transform_publisher-1] translation: ('-0.012000', '0.000000', '0.144000')
[static_transform_publisher-1] rotation: ('0.000000', '0.000000', '0.000000', '1.000000')
[static_transform_publisher-1] from 'base_footprint' to 'laser_frame'
[INFO] [rplidar_composition-2]: process started with pid [8185]
[rplidar_composition-2] [INFO] [1705920254.559700714] [rplidar_composition]: RPLidar running on ROS2 package rplidar_ros. RPLIDAR SDK Version:2.1.0
[rplidar_composition-2] [ERROR] [1705920257.061532757] [rplidar_composition]: Error, operation time out. SL_RESULT_OPERATION_TIMEOUT! 
[ERROR] [rplidar_composition-2]: process has died [pid 8185, exit code 255, cmd '/home/moi/create3_ws/install/rplidar_ros/lib/rplidar_ros/rplidar_composition --ros-args -r __node:=rplidar_composition -r __ns:=/ --params-file /home/moi/create3_ws/install/create3_lidar_slam/share/create3_lidar_slam/config/rplidar_node.yaml'].

```
fails to transform rplidar_compositiion node in rplidar_node
I don't have enough knowledge of ros2 to debug this operation.
Thanks for help!

**Comment by alain-lnpcrd** [comment]
HI,
https://github.com/iRobotEducation/create3_examples/issues/50
![Capture d’écran du 2024-02-08 11-41-13](https://github.com/iRobotEducation/create3_docs/assets/74408456/af7942f7-d6c8-4112-bdd2-6d3a4639af65)
delete https://github.com/iRobotEducation/create3_examples/blob/humble/create3_lidar_slam/launch/sensors_launch.py#L41

**Reply by slowrunner** [reply]
Congratulations on the workaround solution.  What are your plans with the scans?

**Comment by alain-lnpcrd** [comment]
see my create3 wander around the house avoiding prohibited areas such as the indoor garden and possibly make friends with the robot vacuum cleaner.
For now I am starting to configure my aok D lite a1m8 camera and I am entitled to "[ERROR] [component_container-1]: process has died [pid 7936, exit code 127, cmd '/opt/ros/humble/lib/rclcpp_components/component_container --ros-args --log-level info --ros-args -r __node:=oak_container -r __ns:=/']"

**Reply by slowrunner** [reply]
> create3 wander around the house avoiding prohibited areas such as the indoor garden

Looking forward to hearing about your progress and challenges in this - perhaps you can start a "Show and Tell" thread for your robot adventures ([mine](https://github.com/iRobotEducation/create3_docs/discussions/472#discussion-5935028))

> aok D lite a1m8 camera and I am entitled to "[ERROR] [component_container-1]: process has died

Post about this in a new question, and I can probably help with that.  I spent all last week conquering those error msgs.

**Comment by alain-lnpcrd** [comment]
Wonderful your robots!
I separated the board from the create3 by turning on the bluetooth to power the pi with the mains power supply. I manage to get the camera nodes working and I wonder if there is not an insufficient power supply for the camera with the create3 USB; I'll see that tomorrow.
Thank you for your offer of help.
